[POST] Migrer de CJS vers ESM #1 (#38)
1024pix/pix-engineering: _posts/2023-08-25-migrer-de-commonjs-vers-esm.md

---
layout: post
title:  "Migrer de CommonJS vers ECMAScript Modules"
date: 2023-08-25 09:49:00 +0200
categories: javascript
cover:
  image: "cjs-to-esm.png"
  source_url: https://raw.githubusercontent.com/wessberg/cjstoesm/master/documentation/asset/logo.png
authors: pierre_top
---

## ESM, c'est quoi ?

Il y a plusieurs systèmes de [modules](### Module) en Javascript : AMD, CJS, ESM, UMD.
Parmi ceux-ci, le système CommonJs (en abrégé CJS) est [le plus utilisé](https://github.com/wooorm/npm-esm-vs-cjs), car c'était le système utilisé par défaut dans NodeJs, lancé en 2009. A ce moment-là, il n'existait pas de système de module natif, partagé par les navigateurs et NodeJs. 

En 2015, le comité chargé du langage met un terme à cette situation, à savoir :
- le comité TC39 (chargé de Javascript)
- de l'organisation de normalisation ECMA
- publie dans la version 2015

> - la spécification ECMAScript module, en abrégé ESM.

Reviewer: Etrange ces bullet points, c'est voulu?

Author: Oui, sinon la phrase est difficile à comprendre. Si tu as mieux, je prends !

Dès 2017, les navigateurs web et NodeJs l'avaient implémenté. 
En 2020, avec NodeJS 12, le support était complet et il n'était plus considéré comme expérimental.   

## Pourquoi migrer ?

> La base de code Pix, créée en 2011, utilise le format CJS. Si le format qu'elle utilise est éprouvé, pourquoi migrer pour un système qui n'a que trois ans de support officiel ? La décision tient à une raison : faciliter le passage à Typescript.

Reviewer: Le fait que ont passe sur le standard natif ne me semble pas anecdotique

Author: Oui, mais comment justifier ? Le natif c'est bien, mais le standard de facto est CJS, et apprendre ESM est un coût de formation pour les devs. Si on voulait migrer l'API hors NodeJS, vers un runtime sans CJS et juste ESM, ce serait un argument, mais c'est très peu probable.

Le système de module est un sujet complexe, et je ne peux pas le restituer ESM en quelques lignes. Je vais néanmoins lister les avantages de l'expérience de développement : 
- syntaxe moins permissive, qui améliore la lisibilité du code ;
- [mode strict](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Strict_mode) actif ;

> - immutabilité des exports ;

Reviewer: Ca nous a quand même bien complexifier la DX (pour les tests)

Author: Yep, j'en parle dans l'autre PR https://github.com/1024pix/pix-engineering/pull/42
Tu veux que je le mentionne ici ?

- vérification [statique](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/named.md) des imports, absente en CJS.

Pour illustrer la syntaxe, voilà une version valide en CJS.
```javascript
const Ham = require('./ham').Ham;
const carrot = require('./garden').digCarrot();
const potatoes = require('./field').harvestPotatoes(require('./tools').basket);

/* some code */

if( some_condition()){
   require('./orchard').spray();
}

const omelette = cook(require('./egg'));

/* some code */

module.exports = {
   new Ham(),
   require('./spam').someFunction
}
```

Et voilà la version équivalente en ESM :
- les imports sont visibles en haut du fichier, donc plus nombreux ;
- les exports sont évidents. 

```javascript
import { Ham } from './Ham.js';
import egg from './egg.js';
import { someFunction } from './spam.js';
import { digCarrot } from './garden.js';
import { spray } from './orchard.js';

import { harvestPotatoes } from './field';
import { basket } from './tools';
const potatoes = digPotatoes(basket);

const carrot = digCarrot();

/* some code */

if( some_condition() ){
   spray();
}

const omelette = cook(egg);

/* some code */

> const ham = new Ham();
> 
> export {
>    ham,

Reviewer: Mieux vaut être au plus proche du code precedent
```suggestion
export {
   ham: new Ham(),
```

Author: Oui mais en ESM tu peux pas faire ça !

Reviewer: Ah oui, oups!

La syntaxe des exports est la suivante, du plus simple au plus complexe :
  - `export foo;` : exporter un objet, peut être invoqué plusieurs fois
  - `export default foo` : export par défaut (l'objet exporté perd son nom `foo` pour prendre celui de `default`)
  - `export { foo, bar }` : export de plusieurs objets
  - `export { foo , bar as baz }` : même chose, avec renommage
  - `export { foo, bar } from './module.js` : vous avez bien lu, on importe les exports nommés et on les exporte (ré-export)
  - `export * from './module.js` : même chose, pour sélectionner tous les exports nommés

## Outillage

Nous avons besoin de réécrire toute la codebase. Les outils dédiés sont appelées "code modificators" (`codemods`).

https://github.com/facebook/jscodeshift

Il existe un outillage existant incomplet

Besoin de connaître l'utilisation des imports pour modifier les exports

https://github.com/IanVS/eslint-nibble

## Ship faster

Le choix de corriger les syntaxes CJS incompatibles, et de bénéficier des tests automatisés

https://github.com/1024pix/pix/pull/5715
https://github.com/1024pix/pix/pull/5747

## Codemodception

> ![workflow de développement](/assets/images/posts/migrer-de-commonjs-vers-esm/codemod-workflow.png)

Reviewer: Si possible, supprimer le petit texte grisé
<img width="396" alt="image" src="https://github.com/1024pix/pix-engineering/assets/3769147/684a3d88-b782-4583-b0c3-c3253312e406">

Author: Yep, je le ferais dans https://github.com/1024pix/pix-engineering/pull/42

## Le choix des exports nommés

> Sondage dans les équipes

Reviewer: Ce texte est perdu au milieu du reste!?

Reviewer: En fait je ne comprends pas ce chapitre:
"Le choix des exports nommés"
Il semble manquer un liant entre chaque paragraphe

- "Exception" c'est une exception à quel règle?
- Pourquoi le "Ou" alors (ca s'oppose à quoi)

Author: L'article n'était pas fini, ça a atterri dans https://github.com/1024pix/pix-engineering/pull/42

Exception : les fichiers qui n'exportent qu'un seul objet (use-case)
https://github.com/1024pix/pix/pull/5787/files#diff-734269ab23f38145d1f6742bd4282022dc6d47a8f57c3e2b372e45da7f1d88f4

Ou alors on transforme les exports anonymes CSJ en exports nommés CJS (et les imports aussi).

Ca permet de lancer les tests en CJS donc d’aller plus vite.

Et du coup il reste plus que deux codemods CJS=>ESM qui ne gèrent que les imports/exports nommés (simples!).

générer un import nommé ESM (on part du principe que le nom de la variable est prévisible, c’est le nom du fichier snake-case transformé en camelCase/PascalCase)

https://github.com/1024pix/pix/pull/5731

## Le barrelling

## Les tests mockés et l'injection de dépendance

> Le travail d'équipe

Reviewer: A compléter?

Author: L'article n'était pas fini, ça a atterri dans https://github.com/1024pix/pix-engineering/pull/42

## Le suivi et la motivation

https://github.com/1024pix/areweesmyet

![dashboard code](/assets/images/posts/migrer-de-commonjs-vers-esm/are-we-esm-yet.png)

## Le test

https://1024pix.atlassian.net/wiki/spaces/TC1/pages/3754164240/D+ploiement+de+l+API+en+ESM

> On s'est posé la question un peu tard, brainstorming avec les [captains](/organisation/2020/04/14/les-capitaines-de-la-production.html)

Reviewer: Quelle question?

Author: L'article n'était pas fini, ça a atterri dans https://github.com/1024pix/pix-engineering/pull/42

- savoir si ESM [peut t'être utile](#pourquoi-migrer-en-esm-) ;
- comprendre [la syntaxe ESM](#un-peu-de-syntaxe) ;
- comment [modifier le code](#modifier-le-code) (la [syntaxe](#un-peu-de-syntaxe) est un pré-requis).

Si vous connaissez déjà ESM et les AST, attendez plutôt la sortie de l'article [en plat principal](https://github.com/1024pix/pix-engineering/pull/42), à savoir comment nous avons mené la migration.

## Une histoire de modules

Dans la plupart des langages de programmation (C, Java, JavaScript...), le fichier de code (source) est un élément important. Il est implémenté dans le système de fichiers de l'OS, par un fichier texte. Il y a quelque chose de fondamentalement réconfortant à cela. Utiliser un langage qui stocke le code source dans des fichiers aux noms cryptiques et au contenu binaire, c'est une perte d'autonomie.

Pourquoi ? Parce que ranger des fichiers dans des dossiers est une métaphore puissante : elle permet de tracer des frontières dans sa tête, et d'oublier les particularités pour un moment. C'est la même chose qui se passe au niveau en-dessous lorsqu'on crée des fonctions.

> > Aside from the computer itself, the routine is the single greatest invention in computer science. (...) Create a routine to hide information so that you won't need to think about it. (...) Without the abstractive power of routines, complex programs would be impossible to manage intellectually.

Reviewer 2: Juste un HS en passant `We don’t write things down to remember them. We write them down to forget.`
https://www.reproof.app/blog/notes-apps-help-us-forget

Author: Cela s'applique très bien aux journaux personnels maintenant que tu le dis

> in Steve McConnell, Code Complete - Chapter 7 : High-quality routine

Reviewer 2: suggestion : faire une bibliographie à la fin.
exemple :
AUTEUR, Prénom. Titre. Edition. Lieu de publication : éditeur, année. Pagination.
MCCONNELL, Steve. Code Complete. 2nd edition. Washington : Microsoft Press, 2004. 913 p.

Author: Je garde en tête pour mes prochains articles, je vais en faire un template. Merci !

Les imports et exports doivent être à la racine du fichier (`top-level`).
Ils sont interdits dans une structure de contrôle ou une fonction.

Les imports doivent figurer en début de fichier.
Leur syntaxe est la suivante, du plus simple au plus complexe :
  - `import from './module.js'` : si l'on ne veut pas récupérer d'import, uniquement exécuter du code
  - `import module from './module.js'` : pour récupérer l'export par défaut et le nommer `module`
  - `import { foo, bar } from './module.js'` : pour récupérer certains exports nommés
  - `import { foo , bar as baz } from './module.js'` : même chose, en les renommant
  - `import * as module from './module.js'` : pour récupérer tous les exports nommés dans un objet

La syntaxe des exports est la suivante, du plus simple au plus complexe :
  - `export foo;` : exporter un objet, peut être invoqué plusieurs fois
  - `export default foo` : export par défaut (l'objet exporté perd son nom `foo` pour prendre celui de `default`)
  - `export { foo, bar }` : export de plusieurs objets
  - `export { foo , bar as baz }` : même chose, avec renommage

>   - `export { foo, bar } from './module.js` : vous avez bien lu, on importe les exports nommés et on les exporte (ré-export)

Reviewer 2: question : pourquoi avec écrit (ré-export) entre parenthèses et pas on les réexporte ?

Author: Pour expliquer que ré-export (le nom officiel) est ambigü: il ne s'agit pas d'exporter deux fois dans le même fichier

L'affaire se complique si la transformation modifie la syntaxe, et pas seulement un mot-clef.
```javascript
// CJS
const { Ham } = require('./ham');
// ESM
import { Ham } from './Ham.js';
```

Car, même si transformer `const { <TOKEN> }` en  `import { <TOKEN }` semble simple, il faut aussi éviter les effets de bords sur des expressions qui y ressemblent, mais qui n'en sont pas, par exemple les imports nommés non sélectifs.
```javascript
// CJS
const egg = require('./egg.js');
// ESM
import egg from './egg.js';
```

Et que dire si la transformation est multi-lignes ?
```javascript
// CJS
const ham  = new (require('./ham')).Ham();
// ESM
import { Ham } from './Ham.js';
const ham = new Ham();
```

Et encore, ici les lignes sont consécutives, alors qu'il est indispensable en réalité de pouvoir modifier les exports, qui peuvent être n'importe où dans le fichier.

Bref, il nous faut une autre solution : entrons dans le monde de l'AST, dirigé par les `codemods`.

### Une solution complexe et puissante : les codemods

#### L'AST
Les outils dédiés à la modification de code sont appelés "code modificators" (`codemods` en abrégé).

> J'utilise un outil JS,  [jscodeshift](https://github.com/facebook/jscodeshift), comme exemple ci-dessous.

Reviewer 2: `J'utilise un outil JS, [jscodeshift](https://github.com/facebook/jscodeshift), comme exemple ci-dessous.`
Remarque : cette phrase n'est pas hyper claire pour moi, je l'attendrais juste après avoir décris les outils de ce type et juste avant de l'utiliser

Author: Bien vu, je l'ai déplacé

Si on assemble tous les éléments, le codemod ressemble à ça https://github.com/1024pix/pix/pull/5787/commits/1f5f207361cad7a04037429408b2f9545a176e4a

#### Un périmètre de travail parfois limité

Jscodeshift suit les étapes suivantes :
- lire (récursivement) tous les fichiers d'un répertoire ;
- sur chaque fichier :
  - générer l'AST ;
  - appliquer les sélecteurs ;
  - pour chaque sélection qui matche, appeler le constructeur de code ;
  - assembler le code obtenu et réécrire le fichier d'origine.

Le plus grand périmètre de travail est le fichier, tout comme dans eslint.
Même si Jscodeshift traite les fichiers en parallèle, il n'est pas possible d'accéder à l'AST d'un autre fichier depuis l'AST courant. La conséquence : il n'est pas possible de faire une modification qui impacte deux fichiers.

Si les modifications de code ne concernent que l'intérieur d'une fonction, cela ne pose pas de problème.
Dans les cas d'ESM, qui par définition traite des interfaces entre fichiers, cela a été une sérieuse limitation.

### Se faire aider

#### Tests automatisés

Il n'est pas possible, à part avec une profonde expérience du langage et des AST, de se représenter tous les cas possibles.

> Il est donc peu près exclu de garantir que tous les cas seront traités, et qu'ils le soient correctement.

Reviewer 2: suggestion : supprimer la virgule avant le `et` https://www.alloprof.qc.ca/fr/eleves/bv/francais/autres-cas-particuliers-de-la-virgule-f1680

Author: Effectué sur les 11 occurences

Une suite de tests automatisés est donc indispensable pour :
- d'une part, documenter les cas traités ;
- détecter les cas qui ne sont pas bien traités ;
- éviter les régressions, à savoir modifier du code qui ne doit pas être modifié.

Jscodeshift embarque un framework de test dédié, et les cas de test sont lisibles. Ils contiennent le code en entrée (given) et celui attendu en sortie (then).
```javascript
// input.js
const datasource = require('./datasource.js');
// output.js
import * as datasource from './datasource.js';
```

#### Lint

Après l'exécution du codemod, si l'exécution du fichier retourne une erreur, c'est que le codemod est incorrect. Si l'on inspecte le contenu du fichier, on peut constater l'erreur et corriger le codemod. Mais parfois, on a besoin d'une vue d'ensemble pour modifier le codemod, du type : est-ce que tous les imports référencent un fichier existant ?

C'est là que le plugin `eslint-plugin-import` peut faire gagner du temps, par exemple en remontant tous les cas d'imports non résolus avec la règle [no-unresolved](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-unresolved.md).

> Après une exécution de codemod, il peut y avoir beaucoup de violations.

Reviewer 2: ```suggestion
Après une exécution du codemod, il peut y avoir beaucoup de violations.
```

Author: Bien vu. Pour finir, j'ai supprimé ce segment car j'ai utilisé la même formulation plus haut.

Pour y voir plus clair, vous pouvez :
- utiliser le package [eslint-nibble](https://github.com/IanVS/eslint-nibble) qui groupe par règle ;
- exécuter une seule règle sur tous les fichiers avec l'option `--rule` : `npx eslint --rule eslint-plugin-import/no-unresolved ./lib`

> - récupérer les fichier concernés par une règle avec `sed` : `npx eslint ./lib | grep import/named | sed "s/^.*in '\([^']*\)'.*$/\1/"| sed 's/^\(.*\/\)\{1,\}//' | sort | uniq -c | sort`

Reviewer 2: ```suggestion
- récupérer les fichiers concernés par une règle avec `sed` : `npx eslint ./lib | grep import/named | sed "s/^.*in '\([^']*\)'.*$/\1/"| sed 's/^\(.*\/\)\{1,\}//' | sort | uniq -c | sort`
```

Author: Corrigé

---
layout: post
title:  "Migrer de CommonJS vers ECMAScript Modules #1"
date: 2023-08-25 09:49:00 +0200
categories: javascript
excerpt: "Récit d'une aventure : migrer 400kLoc NodeJs en deux mois - #1: les modules"
cover:
  image: "cjs-to-esm.png"
  source_url: https://raw.githubusercontent.com/wessberg/cjstoesm/master/documentation/asset/logo.png
authors: pierre_top
---

## Y'a quoi au menu ?
Cet article partage avec vous deux mois d'efforts (et de doutes, aussi) pour migrer une base de code NodeJS de 400 kLoC. Spoiler : nous y sommes arrivés !

Vous y trouverez de l'histoire, de la syntaxe, des considérations techniques et organisationnelles, de l'outillage et du monitoring : il y en a pour tous les goûts.

Voilà, en entrée, le résumé du challenge à accomplir :
- [comprendre](#une-histoire-de-modules) ce qu'est un système de module ;
- se souvenir [de ce qu'il en est](#les-modules-côté-javascript) en Javascript ;
- savoir si ESM [peut t'être utile](#pourquoi-migrer-en-esm-) ;
- comprendre [la syntaxe ESM](#un-peu-de-syntaxe) ;
- comment [modifier le code](#modifier-le-code) (la [syntaxe](#un-peu-de-syntaxe) est un pré-requis).

> Si vous connaissez déjà ESM et les AST, attendez plutôt la sortie de l'article [en plat principal](https://github.com/1024pix/pix-engineering/pull/42), à savoir comment nous avons mené la migration.

Reviewer 3: Est ce qu'on pourrait écrire en toutes lettres ESM et AST ici, avant d'utiliser les abréviations pour le reste de l'article ?

Author: J'ai revu la phrase pour éviter cela

## Une histoire de modules

Dans la plupart des langages de programmation (C, Java, JavaScript...), le fichier de code (source) est un élément important. Il est implémenté dans le système de fichiers de l'OS, par un fichier texte. Il y a quelque chose de fondamentalement réconfortant à cela. Utiliser un langage qui stocke le code source dans des fichiers aux noms cryptiques et au contenu binaire, c'est une perte d'autonomie.

> Pourquoi ? Parce que ranger des fichiers dans des dossiers est une métaphore puissante : elle permet de tracer des frontières dans sa tête, et d'oublier les particularités pour un moment. C'est la même chose qui se passe au niveau en-dessous lorsqu'on crée des fonctions.

Reviewer: > et d'oublier les particularités pour un moment
Pas certain de comprendre le sens de ça

Author: Je l'ai remplca par 
> elle permet de regrouper ce qui se ressemble en traçant des frontières.

> Aside from the computer itself, the routine is the single greatest invention in computer science. (...) Create a routine to hide information so that you won't need to think about it. (...) Without the abstractive power of routines, complex programs would be impossible to manage intellectually.

in Steve McConnell, Code Complete - Chapter 7 : High-quality routine

En théorie informatique, ce genre de découpage (de décomposition) porte le nom de [module](https://en.wikipedia.org/wiki/Modular_programming), un terme déjà utilisé en [construction](https://www.cnrtl.fr/definition/module) pour exprimer la capacité à bâtir un système en assemblant des composants. La réutilisation de code, sous forme de librairie, en fait partie.

## Les modules côté Javascript

La notion de module est assez large en Javascript: on désigne parfois la même chose par librairie, package, script, balise `<script>`, module, fichier ... Revenons en arrière pour mieux comprendre.

### Tout débute dans un navigateur

> Comme Javascript était initialement (1995) prévu comme un langage limité, dans le navigateur, il n'y avait pas de système de module natif : tout le code était contenu dans la balise `<script></script>`. Au fur et à mesure que les besoins côté navigateur grandissaient (comportement "dynamique"), ils ont été implémentés en Javascript. Comme une partie de ces besoins étaient communs à toutes les équipes (ex : appeler une API REST), les librairies ont été développées. Elles pouvaient être importées depuis un fichier via l'attribut `src` de la balise `script`, par exemple `<script src="my-library.js">`. Le code de l'application (composant) pouvait aussi être modularisé de cette manière.

Reviewer: On pourrait expliciter un peu, et 'javascript' semble superflu ici:
~ils ont été implémentés en Javascript~ ils ont été implémentés mais ne correspondent pas à un standard.

Author: J'ai revu la phrase, qu'en penses-tu ?

### Un petit tour côté serveur

Le passage de Javascript côté serveur (2009) changea l'approche.

Dans NodeJS, les libraires (packages) :
- déclarent leur interface (API) avec le mot-clef `module.exports` (notion de module) ;
- sont enregistrées par l'application dans le fichier `package.json` ;
- sont installées par npm dans le fichier `node_modules` (encore le module) ;
- sont importées avec le mot-clef `require`.

> Les composants de l'application utilisent les mêmes mot-clefs `require` et `module.exports`. Il n'y a pas de différence visible entre l'import d'une libraire et d'un composant, à part le fait qu'une libraire est mentionnée par son nom, et le composant par son chemin.

Reviewer: ```suggestion
Les composants interne de l'application utilisent les mêmes mot-clefs `require` et `module.exports`. Il n'y a pas de différence visible entre l'import d'une libraire et d'un composant, à part le fait qu'une libraire est mentionnée par son nom, et le composant par son chemin.
```

Author: Corrigé

```javascript
// library
const { pick } = require('lodash');
// component
const { controller } = require('./lib/application/certification/certification-controller');
````

### Happy end ?

Maintenant que nous sommes familiers avec le concept de modules dans Javascript, nous arrivons enfin à ESM.

Actuellement, il y a plusieurs systèmes de modules en Javascript : AMD, CJS, ESM, UMD.

Le système CommonJs (en abrégé CJS) est [le plus utilisé](https://github.com/wooorm/npm-esm-vs-cjs) côté serveur, car :
- CJS était fourni avec NodeJs ;
- il n'existait pas de système de module natif, partagé par les navigateurs et NodeJs.

> En 2015, le comité chargé du langage met un terme à cette situation, à savoir :
> - le comité TC39 (chargé de Javascript)
> - de l'organisation de normalisation ECMA
> - publie dans la version 2015
> - la spécification ECMAScript module, en abrégé ESM.

Reviewer: ```suggestion
Le comité [TC39](https://tc39.es/) (chargé du langage Javascript) met un terme à cette situation en publiant dans la version 2015 (ECMAScript 2015/ES6) la spécification ECMAScript module, en abrégé ESM.
```

Author: Oui mais on manque ce que veut dire ECMA. 
Je propose d'aller au plus simple pour trancher
```
ES6 met fin à cette situation en 2015 avec la notion de  ECMAScript Module, en abrégé ESM.
```

> Cinq ans plus tard, en 2020, les navigateurs web et NodeJs (v12) l'avaient implémenté de manière stable.

Reviewer 3: ```suggestion
Cinq ans plus tard, en 2020, les navigateurs web et NodeJs (v12) l'ont implémenté de manière stable.
```

## Pourquoi migrer en ESM ?

> La base de code Pix, créée en 2011, utilise le format CJS. Si le format qu'elle utilise est éprouvé, pourquoi migrer pour un système qui n'a que trois ans de support officiel ? La décision tient à une raison : faciliter le passage à Typescript.

Reviewer 4: Ce n'est pas clair pour moi en quoi utiliser ESM facilite le passage à Typescript.
J'avais cette info en tête quand on a tenté d'utiliser NestJS (dont la doc ne fait référence qu'à Typescript) pour le projet API Data, et qu'on a vu que le projet ne supportait pas ESM (et que ça n'était pas prévu). Et j'avoue que je me suis demandée alors si cette assertion était si évidente.

Author: J'invoque @matthiasferraina

>  CJS vs ESM is not a war, rather a topic to talk even more about, once constraints and respective features are clear, as opposite of just taking a side out of habits, or effort needed, to move on (to ESM).

Je liste tout de même les avantages d'expérience de développement après quelques mois d'utilisation :
- [mode strict](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Strict_mode) actif ;
- autocomplétion des imports par les IDE ;
- vérification [statique](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/named.md) des imports, absente en CJS;
- syntaxe moins permissive, qui améliore la lisibilité du code.

Pour illustrer la syntaxe, voilà une version valide en CJS.
```javascript
const Ham = require('./ham').Ham;
const carrot = require('./garden').digCarrot();
const potatoes = require('./field').harvestPotatoes(require('./tools').basket);

/* some code */

if( some_condition()){
   require('./orchard').spray();
}

const omelette = cook(require('./egg'));

/* some code */

module.exports = {
   new Ham(),
   require('./spam').someFunction

> }

Reviewer: On  peut faire ca?

Reviewer: Comment on va recuperer par exemple l'export `new Ham`?

Author: Très juste, j'ai oublié la clef !

```

Et voilà la version équivalente en ESM :
- les imports sont visibles en haut du fichier ;
- l'interface (= les exports) est claire.

```javascript
import { Ham } from './Ham.js';
import egg from './egg.js';
import { someFunction } from './spam.js';
import { digCarrot } from './garden.js';
import { spray } from './orchard.js';
import { harvestPotatoes } from './field';
import { basket } from './tools';

const potatoes = digPotatoes(basket);
const carrot = digCarrot();

/* some code */

if( some_condition() ){
   spray();
}

const omelette = cook(egg);

/* some code */

const ham = new Ham();

export {
   ham,
   someFunction
}
```

## Un peu de syntaxe

> Pour que vous puissiez suivre la démarche de migration, il est important d'avoir les bases de la syntaxe ESM. Je pars du principe que vous connaissez la syntaxe CJS.

Reviewer 5: Ça pourrait être intéressant de rediriger vers une source pour comprendre la syntaxe CJS. :)

Author: Oui. Après 15 minutes de recherche j'abandonne. 
la doc officielle est pauvre en syntaxe https://nodejs.org/api/modules.html#modules-commonjs-modules
et les autres articles n'abordent pas le sujet, sauf celui-là, mais sans être exhaustif. Je le mets quand même !
https://apprendre-nodejs.fr/v1/chapter-04/index.html#modules

Les imports doivent figurer en début de fichier.
Leur syntaxe est la suivante, du plus simple au plus complexe :
  - `import from './module.js'` : si l'on ne veut pas récupérer d'import, uniquement exécuter du code
  - `import module from './module.js'` : pour récupérer l'export par défaut et le nommer `module`
  - `import { foo, bar } from './module.js'` : pour récupérer certains exports nommés
  - `import { foo , bar as baz } from './module.js'` : même chose, en les renommant
  - `import * as module from './module.js'` : pour récupérer tous les exports nommés dans un objet

La syntaxe des exports est la suivante, du plus simple au plus complexe :
  - `export foo;` : exporter un objet, peut être invoqué plusieurs fois
  - `export default foo` : export par défaut (l'objet exporté perd son nom `foo` pour prendre celui de `default`)
  - `export { foo, bar }` : export de plusieurs objets
  - `export { foo , bar as baz }` : même chose, avec renommage
  - `export { foo, bar } from './module.js` : vous avez bien lu, on importe les exports nommés et on les exporte (ré-export)
  - `export * from './module.js` : même chose, pour sélectionner tous les exports nommés

> Si vous êtes curieux de connaître l'implémentation de ESM, ce [cartoon](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/) est une excellente introduction sur les modules en général et ESM en particulier.

Reviewer: C'est un article avec des "cartoons"
```suggestion
Si vous êtes curieux de connaître l'implémentation de ESM, cet [article](https://hacks.mozilla.org/2018/03/es-modules-a-cartoon-deep-dive/) est une excellente introduction sur les modules en général et ESM en particulier.
```

Author: Corrigé

## Modifier le code

### Une solution simple

> Le passage de CJS à ESM concerne tous les fichiers avec au moins un import / export, autrement dit tous les fichiers.

Reviewer 5: ```suggestion
Le passage de CJS à ESM concerne tous les fichiers avec au moins un import/export, autrement dit tous les fichiers.
```

Author: Corrigé

Étant donné la volumétrie, une modification de code manuelle est exclue. Quelles sont les solutions pour modifier du code ?

Si les modifications sont mineures, par exemple supprimer le `s` de `exports`, les outils quotidiens du développeur suffisent : soit IDE (Find/Replace avec expressions régulières), soit un script bash utilisant find, grep, sed. Si l'on veut éviter d'écrire du bash, une solution est de modifier le fichier avec NodeJs, en utilisant la [librairie I/O](https://nodejs.dev/en/learn/reading-files-with-nodejs/) standard. Les lignes de code sont des chaînes de caractère comme les autres, après tout.

L'affaire se complique si la transformation modifie la syntaxe, et pas seulement un mot-clef.
```javascript
// CJS
const { Ham } = require('./ham');
// ESM
import { Ham } from './Ham.js';
```

> Car, même si transformer `const { <TOKEN> }` en  `import { <TOKEN }` semble simple, il faut aussi éviter les effets de bords sur des expressions qui y ressemblent, mais qui n'en sont pas, par exemple les imports nommés non sélectifs.

Reviewer 5: ```suggestion
Car, même si transformer `const { <TOKEN> }` en  `import { <TOKEN }` semble simple, il faut aussi éviter les effets de bord sur des expressions qui y ressemblent, mais qui n'en sont pas, par exemple les imports nommés non sélectifs.
```

Author: Corrigé

```javascript
// CJS
const egg = require('./egg.js');
// ESM
import egg from './egg.js';
```

Et que dire si la transformation est multi-lignes ?
```javascript
// CJS
const ham  = new (require('./ham')).Ham();
// ESM
import { Ham } from './Ham.js';
const ham = new Ham();
```

Et encore, ici les lignes sont consécutives, alors qu'il est indispensable en réalité de pouvoir modifier les exports, qui peuvent être n'importe où dans le fichier.

Bref, il nous fallait une autre solution. Nous avons cherché comment d'autres avaient fait, car nous ne sommes pas les seuls à vouloir migrer en ESM. Nous avons constaté que leurs solutions ([celle-ci](https://github.com/azu/commonjs-to-es-module-codemod#readme) ou [celle-là](https://github.com/wessberg/cjstoesm)) ne fonctionnaient pas sur notre codebase, mais elles nous ont mis sur la voie.

Entrons dans le monde de l'AST, dirigé par les `codemods`.

### Une solution complexe et puissante : les codemods

#### L'AST
Les outils dédiés à la modification de code sont appelés "code modificators" (`codemods` en abrégé).

> J'utilise un outil JS, [jscodeshift](https://github.com/facebook/jscodeshift), comme exemple ci-dessous.

Reviewer: question: Est-ce ok de passer du `nous` au `je` dans l'article?

Author: Non. J'ai supprimé tous les je :axe:

Ces outils s'appuient sur une représentation du fichier plus riche qu'une suite de chaînes de caractères, pour que nous puissions, par exemple :
- sélectionner un import nommé sélectif CJS ;
- mais pas un import par défaut, ou un import nommé global ;
- et le remplacer par un autre import nommé, cette fois-ci en ESM, ainsi qu'une déclaration de variable.

Pour cela, ils vont effectuer un travail similaire à l'interpréteur NodeJS :
- un lexer qui transforme une chaîne de caractères en "mots" (tokens)
- un parser qui transforme cette suite de tokens en éléments du langage, organisés en arbre.

Cet arbre s'appelle un "Abstract Syntax Tree" (AST).

> #### Sélectionner un élément à modifier

Reviewer 3: ```suggestion
##### Sélectionner un élément à modifier
```

Author: Corrigé

> [Voyons](https://rajasegar.github.io/ast-builder/) ce que donne le fichier contenant un import CJS

Reviewer 3: ```suggestion
Voyons ce que donne le fichier contenant un import CJS dans un [constructeur d'AST](https://rajasegar.github.io/ast-builder/) 
```

Author: Corrigé

        },
        "init": {
          "type": "CallExpression",
          "callee": {
            "type": "Identifier",
            "name": "require"
          },
          "arguments": [
            {
              "type": "Literal",
              "value": "./egg.js",
              "raw": "'./egg.js'"
            }
          ]
        }
      }
    ],
    "kind": "const"
    }
```

Cet "Abstract Syntax Tree" (AST) peut être interrogé, tout comme on interroge le DOM avec Xpath: la requête s'appelle un sélecteur.

Ici, nous voulons récupérer les imports nommés globaux, le sélecteur qui nous intéresse est un nœud :
- avec affectation à une variable (sans invocation de code), donc de type `VariableDeclaration`;
- sans déstructuration, donc avec un descendant de type `Identifier`;
- d'import CJS, donc avec un descendant de type `CallExpression` et de name `require`.

Je n'affiche pas les autres syntaxes d'import (ex : import sélectif) pour la lisibilité, mais il faut tous les comparer pour trouver leurs points communs et leurs différences, pour éviter les faux :
- positifs : sélectionner un import indésirable ;
- négatifs : ne pas sélectionner un import désiré.

> #### Modifier l'élément sélectionné

Reviewer 3: ```suggestion
##### Modifier l'élément sélectionné
```

Author: Corrigé

Si on assemble tous les éléments, le codemod ressemble à ça https://github.com/1024pix/pix/pull/5787/commits/1f5f207361cad7a04037429408b2f9545a176e4a

#### Un périmètre de travail parfois limité

Jscodeshift suit les étapes suivantes :
- lire (récursivement) tous les fichiers d'un répertoire ;
- sur chaque fichier :
  - générer l'AST ;
  - appliquer les sélecteurs ;
  - pour chaque sélection qui matche, appeler le constructeur de code ;
  - assembler le code obtenu et réécrire le fichier d'origine.

Le plus grand périmètre de travail est le fichier, tout comme dans eslint.
Même si Jscodeshift traite les fichiers en parallèle, il n'est pas possible d'accéder à l'AST d'un autre fichier depuis l'AST courant. La conséquence : il n'est pas possible de faire une modification qui impacte deux fichiers.

Si les modifications de code ne concernent que l'intérieur d'une fonction, cela ne pose pas de problème.
Dans les cas d'ESM, qui par définition traite des interfaces entre fichiers, cela a été une sérieuse limitation.

### Se faire aider

#### Tests automatisés

Il n'est pas possible, à part avec une profonde expérience du langage et des AST, de se représenter tous les cas possibles.

> Il est donc peu près exclu de garantir que tous les cas seront traités, et qu'ils le soient correctement.

Reviewer 4: ```suggestion
Il est donc à peu près exclu de garantir que tous les cas seront traités, et qu'ils le soient correctement.
```

Author: Corrigé

Une suite de tests automatisés est donc indispensable pour :
- d'une part, documenter les cas traités ;
- détecter les cas qui ne sont pas bien traités ;
- éviter les régressions, à savoir modifier du code qui ne doit pas être modifié.

Jscodeshift embarque un framework de test dédié, et les cas de test sont lisibles. Ils contiennent le code en entrée (given) et celui attendu en sortie (then).
```javascript
// input.js
const datasource = require('./datasource.js');
// output.js
import * as datasource from './datasource.js';
```

#### Lint

> Après l'exécution du codemod, si l'exécution du fichier retourne une erreur, c'est que le codemod est incorrect. Si l'on inspecte le contenu du fichier, on peut constater l'erreur et corriger le codemod. Mais parfois, on a besoin d'une vue d'ensemble pour modifier le codemod, du type : est-ce que tous les imports référencent un fichier existant ?

Reviewer 3: ```suggestion
Après l'exécution du codemod, si l'exécution du fichier retourne une erreur, c'est que le codemod est incorrect. Si l'on inspecte le contenu du fichier, on peut constater l'erreur et corriger le codemod. Mais parfois, on a besoin d'une vue d'ensemble pour modifier le codemod, par exemple pour s'assurer que tous les imports référencent un fichier existant.
```

Author: Corrigé

C'est le cas de l'import nommé.
```javascript
// CJS
const egg = require('./egg.js');
// ESM
import egg from './egg.js';
```

##### Cas 2 - Questionnement : faut-il automatiser ?
Si une fonctionnalité CJS n'est pas reprise dans ESM, et est facilement simulable (ex : avec plusieurs instructions au lieu d'une), alors se demander si elle est utilisée dans la plupart de la codebase.

Si c'est le cas, écrire le codemod. Si ce n'est pas le cas, modifier les fichiers manuellement.

C'est le cas CJS de l'import + invocation dans la même instruction, remplacée par un import nommé avec assignation + invocation dans une autre instruction en ESM.

```javascript
// CJS
const ham  = new (require('./ham')).Ham();
// ESM
import { Ham } from './Ham.js';
const ham = new Ham();
```

##### Cas 3 - Impasse : il est impossible d'automatiser
Si une fonctionnalité CJS n'est pas reprise dans ESM, et qu'elle n'est pas simulable facilement, la seule solution est de modifier le fichier manuellement.

> C'est le cas des [exports immuables](https://2ality.com/2015/07/es6-module-exports.html) en ESM. Tous les tests CJS qui substituent une [doublure de test](https://martinfowler.com/bliki/TestDouble.html) à la volée ne sont plus valables : seule [l'injection des dépendances](https://github.com/1024pix/pix/commit/027b560b42ef72731662559f906de33ff749a008) permet de résoudre le problème. Bien sûr, il est possible en théorie d'écrire un codemod qui injecterait les dépendances : à vous de voir le coût.

Reviewer: ```suggestion
C'est le cas des [exports immuables](https://2ality.com/2015/07/es6-module-exports.html) en ESM. Tous les tests CJS qui substituent une [doublure de test](https://martinfowler.com/bliki/TestDouble.html) (stubs, mocks etc...) à la volée ne sont plus valables : seule [l'injection des dépendances](https://github.com/1024pix/pix/commit/027b560b42ef72731662559f906de33ff749a008) permet de résoudre le problème. Bien sûr, il est possible en théorie d'écrire un codemod qui injecterait les dépendances : à vous de voir le coût.
```

Author: Je préfère faire le snob pour éclaircir la situation !
